library/core/qsl-base/src/main/java/org/quiltmc/qsl/api/event/ParameterInvokingEvent.java

```diff
@@ -0,0 +1,79 @@
+ * {@code ScaryEvents.Spooked}. {@code Spooked} is a functional interface where the first parameter is a ghost and the
+ * second parameter is the entity being spooked. The entity being spooked is the invoking parameter.
+ *
+ * <h2>Why make an event parameter invoking?</h2>
+ *
+ * A good reason to make an event parameter invoking is when the event directly relates to an action performed on a
+ * specific object. For example, an entity being unloaded from a world. With a traditional event, you would register a
+ * callback and check for your specific entity in the callback, typically via an {@code instanceof} check for modded
+ * entities. With a parameter invoking event, if you directly control the object the action is being performed on such
+ * as the object being one of your own modded entities, you can implement the callback interface, {@code T}, and have
+ * the event execute the callback on your own object, removing the need to find your specific entity through further
+ * checks.
+ *
+ * <p>Assuming {@code ScaryEvents.SPOOKED}'s callback interface is {@code Spooked} and the event is parameter invoking,
+ * you may implement {@code ScaryEvents.Spooked} directly on to your Entity to be notified that when your entity is
+ * spooked.
+ *
+ * <pre><code>
+ * class MyEntity extends MobEntity implements ScaryEvents.Spooked {
+ *     ...
+ *
+ *     &#64;Override
+ *     public void spooked(Ghost ghost, Entity ignored) {
+ *         // This method will be invoked when the event, `ScaryEvents.SPOOKED` is executed and the invokable parameter
+ *         // is instanceof MyEntity
+ *
+ *         // My entity is very scared of ghosts, hence it should die when spooked.
+ *         this.setHealth(0);
+ *     }
+ * }
+ * </code></pre>
+ *
+ * <h2>How to use this annotation</h2>
+ *
+ * An event which may invoke a parameter should first place the annotation on the field of the event instance or the
+ * method which returns an event instance.
+ *
+ * <pre><code>
+ * &#64;ParameterInvokingEvent
+ * public static final Event&lt;ScaryEvents.Spooked&gt; SPOOKED = EventFactory.createArrayEvent(ScaryEvents.Spooked.class, callbacks -> (ghost, spookedEntity) -> {
+ *     // Since this event is parameter invoking, does the entity being spooked implement `Spooked`?
+ *     if (spookedEntity instanceof ScaryEvents.Spooked callback) {
+ *         // Since the entity implements the callback, invoke the callback on the entity.
+ *         // For this event we invoke on the parameter first, other events may do this differently.
+ *         callback.spooked(ghost, spookedEntity);
+ *     }
+ *
+ *     for (var callback : callbacks) {
+ *     	   // Invoke the callbacks implemented by this event
+ *         callback.spooked(ghost, spookedEntity);
+ *     }
+ * });
+ * </code></pre>
+ */
+@Documented // Documentation
+@Retention(RetentionPolicy.CLASS) // For static analysis
+@Target({ ElementType.FIELD, ElementType.METHOD })
+public @interface ParameterInvokingEvent {
```

Comment:
That annotation name is still super obscure, at least the doc is clear.

Reply:
Do you have a better idea?

Reply:
Sadly, no.
I don't think this can be named in a clear way anyway, and this is the best name so far.

Reply:
`ParameterImplementedEvent` would be more clear to me, but at this point the documentation is good enough at describing its function that the name probably doesn't matter 😄

Reply:
Ah the age old naming question.

We could go for a more Gradle-like naming scheme with something like `InvokesParameterAsCallback` for a name.